tokio/src/time/interval.rs

```diff
@@ -103,26 +109,179 @@ pub fn interval_at(start: Instant, period: Duration) -> Interval {
     Interval {
         delay: Box::pin(sleep_until(start)),
         period,
+        missed_tick_behavior: Default::default(),
     }
 }
 
-/// Interval returned by [`interval`](interval) and [`interval_at`](interval_at).
+/// Defines the behavior of an [`Interval`] when it misses a tick.
+///
+/// Sometimes, an [`Interval`]'s tick is missed. For example, consider the
+/// following:
+///
+/// ```ignore
+/// use tokio::time;
+///
+/// let mut interval = time::interval(time::Duration::from_secs(2));
+/// for _ in 0..5 {
+///     interval.tick().await;
+///     // if this takes more than 2 seconds, a tick will be delayed
+///     task_that_takes_one_to_three_seconds().await;
+/// }
+/// ```
+///
+/// Generally, a tick is missed if someone spends too much time without calling
+/// [`tick()`](Interval::tick).
+///
+/// By default, when at tick is missed, [`Interval`] fires ticks as quickly as
+/// it can until it is back to where it should be. However, this is not always
+/// the desired behavior. `MissedTickBehavior` allows users to specify which
+/// behavior they want [`Interval`] to exhibit. Each variant represents a
+/// different strategy.
+///
+/// Note that because the executor cannot guarantee exect precision with timers,
+/// these strategies will only apply when the error in time is greater than 5
+/// milliseconds.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum MissedTickBehavior {
+    /// Tick as fast as possible until caught up.
+    ///
+    /// When this strategy is used, [`Interval`] schedules ticks "normally" (the
+    /// same as it would have if the ticks hadn't been delayed), which results
+    /// in ticks being fired as fast as it can until it is caught up in time to
+    /// where it should be.
+    ///
+    /// This looks something like this:
+    /// ```text
+    /// Expected ticks: |     1     |     2     |     3     |     4     |     5     |     6     |
+    /// Actual ticks:   | work -----|          delay          | work | work | work -| work -----|
+    //  Poll behavior:  |   |       |                         |      |      |       |           |
+    //                  |   |       |                         |      |      |       |           |
+    //           Ready(s)   |       |             Ready(s + 2p)      |      |       |           |
+    //                Pending       |                    Ready(s + 3p)      |       |           |
+    //                   Ready(s + p)                           Ready(s + 4p)       |           |
+    //                                                                  Ready(s + 5p)           |
+    //                                                                              Ready(s + 6p)
+    // * Where `s` is the start time and `p` is the period
+    /// ```
```

Comment:
Does the new commit help to simplify? The poll behavior stuff was to help me figure out how to implement it. I left it in there for documentation purposes, but it seems to add more confusion. I used the same type of diagram that @Lindenk used in #3574. The idea is that 'work' is how long we spend doing, well, actual work, the dashes are time we spend (asynchronously) waiting for the tick, and 'delay' is representative of a situation where an `Interval` would be delayed from ticking.

So, we would expect the ticks to proceed along like diagrammed in the first line ('Expected ticks:'). But if the `Interval` is delayed, the bottom line ('Actual ticks:') shows how the `Interval` would behave. In this specific case, notice that there is no delay (no dashes) in the firing of ticks after the long 'delay' tick. That's because the specific behavior this diagram represents doesn't wait for ticks until it is "caught up" in time to where it should be. In this case, after 1 normal tick, one excessively long tick, and 2 ticks done without any delay, the 5th tick starts late, but finishes its work before the next tick was supposed to fire, so it waits for a little bit before tick 6 fires.